### DISCUSSION TITLE
Support for USB RNDIS network devices

### DISCUSSION THREAD

**Discussion by lars18th** [antonym/netboot.xyz#600] [General] [answered]
Hi,

I'm testing some of the LiveCD assets with **only** USB network devices. And I found that a lot of them doesn't found any network device when booting. Mainly all Ubuntu LiveCD assets fail in this point.

Please, can you update the scripts that generate the initram assets to include USB nework devices?

Thank you!

**Comment by thelamer** [comment]
We don't add things to their initramfs outside of overlaying the bare necessities (portable curl bin and script mods to support multipart files)

You are talking about ingesting kernel modules and such from their repos on a rolling basis and embedding them into not only the initramfs, but their squashfs. 
Then on top of that you need to modify all of their upstream repos pre-init boot logic to support USB devices (they conventionally have different names and are not supported by networking init).

Two things here: 
1. I do not think you understand the scope of something like this given the amount of distros we ingest
2. You really should be reaching out upstream to individual projects to have them include better network support in their Live assets.

**Comment by lars18th** [comment]
Hi @thelamer ,

Yes, I'm a newcomer to this project, and I don't understand much about the scope of it. I'm still testing and reading. However, I want to report the troubles that I see. The lack of sufficient USB nic support is now a problem. And more and more portable devices doesn't have PCI nics. So I feel it's a real problem. Regarding the part of the ipxe support, I writed about it in another issue (#599). But here I want to comment about solutions to this problem. If the best solution is to comment directly with the LiveCD releasers, then I'll try to do it. But, in any case, I suggest to document the list of distributions/tools with USB device support. For instance, a simple list like:
```
+-------------------------------------+-------+-------+
| Distribution                        |  PCI  |  USB  |
+-------------------------------------+-------+-------+
+ Ubuntu XX.XX                        +  YES  +  no   +
+ AVG Live CD                         +  YES  +  YES  +
|   ...                                               |
+-------------------------------------+-------+-------+
```

Please, consider it.
Regards.

**Comment by thelamer** [comment]
It is just the nature of the beast, same with WIFI. 
Have you been able to identify any distros that support USB devices ? 
Even Dracut/Fedora which is basically the shining star in pre-init and network support does not include USB network device driver modules in their initramfs.

**Comment by lars18th** [comment]
Hi @thelamer ,

> It is just the nature of the beast, same with WIFI.

Good point! Then perhaps the best is to document ETH, USB and WIFI nics.
You agree?

> Have you been able to identify any distros that support USB devices ?

I'll try to list _some_ here.

> Even Dracut/Fedora which is basically the shining star in pre-init and network support does not include USB network device driver modules in their initramfs.

So it will be required to contact with the authors. I understand that with a ISO/CDROM or USB disk the USB network modules are accessible after the initial kernel boot. However, when using initram+squashfs this is not true. The develpers require to note this deficiency and add the corresponding support.

Regards.

**Comment by lars18th** [comment]
Hi @thelamer ,

Here some non-extensive tests:
```
Network NIC driver support based on bus type:

+--------------------------------------------------+-------+-------+-------+
| Distribution                                     |  PCI  | WIFI  |  USB  |
+--------------------------------------------------+-------+-------+-------+
| Linux Install -> Arch Linux 2020                 |  YES  |  ??   |  YES  |
| Linux Install -> Debian 10 graphic               |  YES  |  ??   |  YES* |
| Linux Install -> Fedora 31 Server                |  YES  |  ??   |  no   |
| Linux Install -> Ubuntu 19.10                    |  YES  |  ??   |  YES  |
|                                                  |       |       |       |
| LiveCD -> Debian Live 10 (cinnamon)              |  YES  |  ??   |  no   |
| LiveCD -> Fedora Live 31 Xfce                    |  YES  |  ??   |  no   |
| LiveCD -> Linux Lite 4                           |  YES  |  ??   |  no   |
| LiveCD -> LXLE                                   |  YES  |  ??   |  no   |
| LiveCD -> Tiny Core Linux 64 (pure)              |  YES  |  ??   |  no   |
| LiveCD -> SparkyLinux Live Xfce                  |  YES  |  ??   |  no   |
| LiveCD -> Voyager Live Bionic                    |  YES  |  ??   |  no   |
|                                                  |       |       |       |
| Utilities -> AVG Rescue CD                       |  YES  |  ??   |  YES  |
| Utilities -> Boot Repair CD                      |  YES  |  ??   |  no   |
| Utilities -> CAINE                               |  YES  |  ??   |  no   |
| Utilities -> Clonezilla Debian                   |  YES  |  ??   |  YES  |
| Utilities -> Clonezilla Ubuntu                   |  YES  |  ??   |  no   |
|                                                  |       |       |       |
+--------------------------------------------------+-------+-------+-------+

*: Manual selection of the driver
```

Perhaps you want to add this (or similar) table to the documentation and add updates.

I hope it helps.
Regards.

**Comment by lars18th** [comment]
Hi @thelamer ,

Any news about improve the support for USB network devices (Ethernet and Wifi)?

Regards.

**Comment by thelamer** [comment]
I have participated heavily in the project for loading new assets, but my knowledge of core C stuff for IPXE is limited. 
Ultimately this decision to add more drivers to the IPXE build would land in @antonym 's lap as he has a better understanding of any unintended consequences like filesize limitations on platforms.

**Comment by lars18th** [comment]
Hi @thelamer ,

I feel they are two different things to improve: the support of ipxe for USB devices, and the USB support for distributions. Your agree with that?

**Comment by antonym** [comment]
Support for iPXE USB/Wifi would need to be raised, added, or improved at the upstream iPXE project.  The current support there for those items is rather limited.  

Adding support for USB within the distributions is another upstream task that probably wouldn't belong here.  We make best effort to make things work with distributions, but ultimately it is up to each individual distribution as to whether they even want to support us or not.  If support is added upstream to either component, it will naturally filter down here and we will be able to consume that support.

**Comment by lars18th** [comment]
Hi @antonym ,

Thank you for the comment!
Then I can assume that two things are involved.

So regarding the "distribution support":

- Even this project can't do much to improve the support, it will be desirable to know which distributions have support for USB Ethernet and USB Wifi. You agree with that?

And regarding the "iPXE USB/Wifi support":

- Here I feel two actions can be done. The first can be to contact with the upstream developers to improve this support. And I don't know how to do this as the current official forum seems dead and impossible to sign it. And the second is to prepare some support inside netbootxyz. For example, at time is possible to compile additional USB drivers with the iPXE binary. Perhaps a first step can be to prepare an alternative binary. What you think?

**Comment by samip5** [comment]
My take on this is just that it's not that easy to do as you seem to make it sound like, there's limitations on how big the loading rom, in this case iPXE can be, and adding USB/WiFI support to it is a C coding problem which can only be fixed or done something about upstream (ipxe.org) and probably time intensive as well. I don't think this is really in scope of the project but I might be wrong so correct me Antony, if I indeed am wrong.

WiFi for one would be really hard to support due to all of the different chipsets which are supported by different version of Linux kernel and it does not usually even support PXE booting to begin with.

**Comment by lars18th** [comment]
Hi,

Instead of adding a complexity that is not required for this project, a simple request:

Will be at any time some support for devices without Ethernet ports?

Please, note that today all devices with USB ports can use Ethernet adapters, and the boot can be chainloaded from a simple USB drive. So my request are for these devices... that more and more are created by the industry (personally I hate notebooks without Ethernet, but they start to be very common!).

Regards.

**Comment by Stromduster** [comment]
Is there a possibility to add this feature ? A lot of laptops come without ethernet, supporting them with ethernet USB adapter or Wifi embedded card would be great.

**Reply by samip5** [reply]
Not really as ipxe has very limited support for doing that.

**Reply by lars18th** [reply]
> Not really as ipxe has very limited support for doing that.

Why do you say that? RNDIS support has been in ipxe for years! Enabling it in netbootxyz is straightforward.

**Reply by samip5** [reply]
> > Not really as ipxe has very limited support for doing that.
> 
> 
> 
> Why do you say that? RNDIS support has been in ipxe for years! Enabling it in netbootxyz is straightforward.
> 
> 

Because even if you enable it, the likelyhood of it causing more support issues is high due to not every imaginable USB Ethernet/Wifi will work.

**Reply by lars18th** [reply]
> > Why do you say that? RNDIS support has been in ipxe for years! Enabling it in netbootxyz is straightforward.
> 
> Because even if you enable it, the likelyhood of it causing more support issues is high due to not every imaginable USB Ethernet/Wifi will work.

Please, read about RNDIS. This is an _universal_ protocol to access to NIC **Ethernet** interfaces, like USB HID for Keyboard or Mouse. Today any imaginable USB Ethernet adapter that supports it will work. The problem is not ipxe, I'm using RNDIS with success. The problem is the netbootxyz support for it. And please, don't confuse WIFI (out-of-scope) with RNDIS Ethernet.

**Reply by samip5** [reply, marked as answer]
> > > Why do you say that? RNDIS support has been in ipxe for years! Enabling it in netbootxyz is straightforward.
> > 
> > 
> > Because even if you enable it, the likelyhood of it causing more support issues is high due to not every imaginable USB Ethernet/Wifi will work.
> 
> Please, read about RNDIS. This is an _universal_ protocol to access to NIC **Ethernet** interfaces, like USB HID for Keyboard or Mouse. Today any imaginable USB Ethernet adapter that supports it will work. The problem is not ipxe, I'm using RNDIS with success. The problem is the netbootxyz support for it. And please, don't confuse WIFI (out-of-scope) with RNDIS Ethernet.

Original point of my comment was the original ` or Wifi embedded card`.
